### ISSUE TITLE
Compile warnings using LLVM/Clang 9.0.0 / C++17 with Visual Studio 2019

### ISSUE THREAD

**Issue by Lunatix89** [gabime/spdlog#1109]
Including FMT and compiling with C++17 standard yields the following compiler deprecation warning: 

```
In file included from ..\..\lib\spdlog\include\spdlog\spdlog.h:12:
  In file included from ..\..\lib\spdlog\include\spdlog/common.h:30:
  In file included from ..\..\lib\spdlog\include\spdlog/fmt/fmt.h:22:
D:\lib\spdlog\include\spdlog\fmt\bundled\core.h(202,25): warning : 'result_of<fmt::v5::internal::custom_formatter<wchar_t, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t> > (int)>' is deprecated: warning STL4014: std::result_of and std::result_of_t are deprecated in C++17. They are superseded by std::invoke_result and std::invoke_result_t. You can define _SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to acknowledge that you have received this warning. [-Wdeprecated-declarations]
    typedef typename std::result_of<
                          ^
  ..\..\lib\spdlog\include\spdlog/fmt/bundled/core.h(830,34):  note: in instantiation of template class 'fmt::v5::internal::result_of<fmt::v5::internal::custom_formatter<wchar_t, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t> > &(int)>' requested here
  FMT_CONSTEXPR typename internal::result_of<Visitor(int)>::type
                                   ^
  ..\..\lib\spdlog\include\spdlog/fmt/bundled/format.h(3096,10):  note: while substituting deduced template arguments into function template 'visit_format_arg' [with Visitor = fmt::v5::internal::custom_formatter<wchar_t, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t> > &, Context = fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t>]
      if (!visit_format_arg(f, arg))
           ^
  ..\..\lib\spdlog\include\spdlog/fmt/bundled/format.h(2031,15):  note: in instantiation of member function 'fmt::v5::format_handler<fmt::v5::arg_formatter<fmt::v5::back_insert_range<fmt::v5::internal::basic_buffer<wchar_t> > >, wchar_t, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t> >::on_replacement_field' requested here
        handler.on_replacement_field(p);
                ^
  ..\..\lib\spdlog\include\spdlog/fmt/bundled/format.h(3237,10):  note: in instantiation of function template specialization 'fmt::v5::vformat_to<fmt::v5::arg_formatter<fmt::v5::back_insert_range<fmt::v5::internal::basic_buffer<wchar_t> > >, wchar_t, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t> >' requested here
    return vformat_to<arg_formatter<range>>(
           ^
  ..\..\lib\spdlog\include\spdlog/fmt/bundled/format-inl.h(954,13):  note: in instantiation of function template specialization 'fmt::v5::internal::vformat_to<wchar_t>' requested here
    internal::vformat_to(buffer, format_str, args);
              ^
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.20.27508\include\type_traits(1529,8):  note: 'result_of<fmt::v5::internal::custom_formatter<wchar_t, fmt::v5::basic_format_context<std::back_insert_iterator<fmt::v5::internal::basic_buffer<wchar_t> >, wchar_t> > (int)>' has been explicitly marked deprecated here
  struct _CXX17_DEPRECATE_RESULT_OF result_of { // explain usage
         ^
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.20.27508\include\yvals_core.h(628,7):  note: expanded from macro '_CXX17_DEPRECATE_RESULT_OF'
      [[deprecated("warning STL4014: "                                                    \
        ^
```
The feature removal information can be found here (search for _SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING): https://devblogs.microsoft.com/cppblog/c17-feature-removals-and-deprecations/

A workaround is to define "_SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING" to silence the warning.

Clang output from "clang --version":
```
clang version 9.0.0 (trunk)
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
```

I also opened a new issue in the FMT repository for this: https://github.com/fmtlib/fmt/issues/1198

**Comment by gabime** [comment]
Thanks for reporting. Will wait for fmt fix

**Comment by Lunatix89** [comment]
Issue in fmt has been closed because it has already been fixed in master as https://github.com/fmtlib/fmt/commits?author=vitaut statet statet.

See commit:
https://github.com/fmtlib/fmt/commit/406e632bd1d956450ae1690d9b5231c0225f91ac

I think users compiling with C++17 will have to define `_SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING` and wait for the next release of both, fmt and spdlog :)

**Comment by gabime** [comment]
We could add a ```pragma disable(...)``` around the include of fmt.h the meantime. Whats the number ?   4014?

**Comment by Lunatix89** [comment]
Yes.. but it seems that clang does not understand the `#pragma warning disable` directive like MSVC does.
There seems to be only the option to disable all deprecation warnings :|

```c++
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
#elif defined(_MSC_VER)
#pragma warning(push)
// this is untested - I don't know whether the "STL" prefix is required or not
#pragma warning (disable: STL4014) 
#endif

#include "fmt/fmt.h"

#if defined(__clang__)
# pragma clang diagnostic pop
#elif defined(_MSC_VER)
# pragma warning(pop)
#endif
```

**Comment by gabime** [comment]
> Is it really necessary to put such.. hacks? into your codebase just to silence such warnings?

I agree. Better not silence warnings in general. 
This will wait for official fmt release.

**Comment by Lunatix89** [comment]
Thank you for spending your time on this :)

And thanks for this awesome library 😃 
I especially like the colorized output and the `msvc_sinc` 😉

**Comment by gabime** [comment]
Thanks for the nice feedback!

**Comment by gabime** [comment]
Fixed in pull #1213
